presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -211,7 +253,62 @@ public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph
                     joinFilters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(joinFilters)),
                     Optional.empty(),
                     Optional.empty(),
-                    Optional.empty());
+                    Optional.empty()));
+        }
+
+        private boolean introducesCrossJoin(List<PlanNode> leftSources, List<PlanNode> rightSources)
```

Comment:
I'm not sure of a cheaper way to do this.  This should generally still be better than considering all those plans, but might cause problems if we have a very connected join graph so you don't end up eliminating anything.  We'll be doing benchmarking of the time and resource usage on the coordinator for join enumeration.